### ISSUE TITLE
The page isn't redirecting properly

### ISSUE THREAD

**Issue by Hugao** [qooob/authentic-theme#247]
After updating to version 16, after the login gives the error *The page isn't redirecting properly*.
Both Firefox and Midori give the error.

I'm using CentOS 7

**Comment by qooob** [comment]
You should hard rest your browsers cache and restart Webmin.

**Comment by Hugao** [comment]
@qooob already did it. It says to many redirects

**Comment by qooob** [comment]
What about putting cursor to URL line in Firefox and hitting enter? That shouldn't be the theme's case really..

**Comment by Hugao** [comment]
@qooob since was on a virtual machine, I recover from a snapshot without webmin and after the update of the theme the problem happened again

**Comment by qooob** [comment]
Could you please try Chrome browser? Did you also try rebooting your machine?

In case it happens in Chrome, please make a screenshot from Chrome with DevTools opened.. O really have no idea why this could be happening.

Try checking
http://www.howtogeek.com/forum/topic/firefox-error-the-page-isnt-redirecting-properly
https://support.mozilla.org/en-US/questions/985967

**Comment by zamolxe** [comment]
I can confirm the same issue
Running CentOS7
Tested using Firefox and Safari on Mac (cleaned cache and cookies)

<img width="1049" alt="screen shot 2015-09-19 at 13 45 32" src="https://cloud.githubusercontent.com/assets/1435948/9976054/387bf536-5ed5-11e5-8d74-9f908d205829.png">

**Comment by qooob** [comment]
There is also Webmin cache. It could get stuck. Clear server side cache as well.

https://support.mozilla.org/en-US/questions/938531

What is your Webmin version, guys?

**Comment by zamolxe** [comment]
I cleaned everything that could be cleaned, same behaviour.
Webmin 1.760.
Below is a snapshot of the webmin debug log (all logging enable), and the log goes on with same entries until firefox/safari stops trying to load the page.

4719 [19/Sep/2015 13:38:29.004928] - 10.10.0.58 - START "script=index.cgi"
4719 [19/Sep/2015 13:38:29.005235] - 10.10.0.58 - READ "/etc/webmin/miniserv.conf"
4719 [19/Sep/2015 13:38:29.005738] root 10.10.0.58 - READ "/usr/libexec/webmin/authentic-theme/config"
4719 [19/Sep/2015 13:38:29.007453] root 10.10.0.58 - READ "/usr/libexec/webmin/lang/en.UTF-8"
4719 [19/Sep/2015 13:38:29.008967] root 10.10.0.58 - READ "/etc/webmin/custom-lang"
4719 [19/Sep/2015 13:38:29.009169] root 10.10.0.58 - READ "/etc/webmin/custom-lang.en.UTF-8"
4719 [19/Sep/2015 13:38:29.018060] root 10.10.0.58 - READ "/usr/libexec/webmin/authentic-theme/lang/en.UTF-8"
4719 [19/Sep/2015 13:38:29.019280] root 10.10.0.58 - READ "/etc/webmin/authentic-theme/custom-lang"
4719 [19/Sep/2015 13:38:29.019509] root 10.10.0.58 - READ "/etc/webmin/authentic-theme/custom-lang.en.UTF-8"
4719 [19/Sep/2015 13:38:29.020702] root 10.10.0.58 - READ "/usr/libexec/webmin//defaultacl"
4719 [19/Sep/2015 13:38:29.020952] root 10.10.0.58 - READ "/etc/webmin/root.acl"
4719 [19/Sep/2015 13:38:29.021193] root 10.10.0.58 - READ "/etc/webmin//root.acl"
4719 [19/Sep/2015 13:38:29.021422] root 10.10.0.58 - READ "/etc/sysconfig/network"
4719 [19/Sep/2015 13:38:29.021788] root 10.10.0.58 - CMD "cmd=hostname"
4719 [19/Sep/2015 13:38:29.025883] root 10.10.0.58 - READ "/usr/libexec/webmin/virtual-server/module.info"
4719 [19/Sep/2015 13:38:29.073748] root 10.10.0.58 - READ "/usr/libexec/webmin/server-manager/module.info"
4719 [19/Sep/2015 13:38:29.085647] root 10.10.0.58 - READ "/etc/webmin/module.infos.cache"
4719 [19/Sep/2015 13:38:29.108561] root 10.10.0.58 - READ "/etc/webmin/installed.cache"
4719 [19/Sep/2015 13:38:29.119455] root 10.10.0.58 - READ "/usr/libexec/webmin/acl/module.info"
4719 [19/Sep/2015 13:38:29.119916] root 10.10.0.58 - READ "/etc/webmin/webmin.cats"
4719 [19/Sep/2015 13:38:29.120124] root 10.10.0.58 - READ "/etc/webmin/webmin.descs"
4719 [19/Sep/2015 13:38:29.120768] root 10.10.0.58 - READ "/usr/libexec/webmin/change-user/module.info"
4719 [19/Sep/2015 13:38:29.121586] root 10.10.0.58 - READ "/etc/webmin/authentic-theme/favorites.json"
4719 [19/Sep/2015 13:38:29.132471] root 10.10.0.58 - READ "/usr/libexec/webmin//module.info"
4719 [19/Sep/2015 13:38:29.133591] root 10.10.0.58 - STOP "runtime=0"

**Comment by qooob** [comment]
Can you manually downgrade to let's say version 14 to see what will happen then?
https://github.com/qooob/authentic-theme/raw/master/.build/authentic-theme-14.00.wbt.gz

I guess you could manually remove Webmin session by deleting `/var/webmin/sessiondb.pag`

I don't see how this issue is related to the theme.

**Comment by zamolxe** [comment]
I downgraded to version 14. Works without any problems as before.

**Comment by qooob** [comment]
Can you also clear the cookies? If not just run it in incognito mode, click `Ctrl+Shift+N` (for Chrome), does it work then?

**Comment by zamolxe** [comment]
Tried the incognito mode
v14: works
v16: same behaviour, redirect loop

**Comment by qooob** [comment]
Reply from [Firefox Forums](https://support.mozilla.org/sv/questions/938531):

> To solve this problem there is no need to delete all of your cookies, simply delete the ones associated with your google and youtube accounts. Here's how:

> - Click Tools/Options/Privacy

> - Click "Remove Individual Cookies"

> - In the Cookies Window type "accounts" (no quotes) into the Search box

> - Select all the cookie names starting with:  accounts.google.com  and  accounts.youtube.com

> - Click Remove Cookie

> You should now be able to reach your Gmail or Youtube login page as usual.
This is how I was able to solve my problem. Deleting other google.com and youtube.com cookies did not help, only after I deleted the account cookies I was able to get back to the login page.

**Comment by qooob** [comment]
I tried different version of Firefox in Linux, OS X, Windows - I can not reproduce the bug.

Please be kind making video screencast for this process of signin in, while keeping DevTools opened. I have no clue why this is happening and I have very strong certainty that it's _not_ theme related!

**Comment by zamolxe** [comment]
I hope your premonition is right :) 
Anyhow, I did a screen cast so you can see what I see, below is the link
https://youtu.be/c7mjJFtxY9g

**Comment by andreychek** [comment]
I've been doing some testing on various browsers, and so far I haven't been able to reproduce this issue.

I'll continue reviewing the issue, though I may ask Jamie to see if he has any input.

**Comment by qooob** [comment]
Ok, I think I know where the problem lies! You shouldn't use that module to change the theme. 

Please goto `Webmin->Webmin Configuration->Themes`. After you set Authentic Theme globally and see nice login (Authentic login screen) - you will be fine! Try it and let me know. Make sure that in `Webmin->Change Language and Theme` you **have default** theme set first. Then do procedure above!

Helped?

**Comment by zamolxe** [comment]
Well... Tried the above, exactly as described by you.
Now I get the nice looking login, but after the login goes again in the loop.

**Comment by qooob** [comment]
That should not be happening! :|

**Comment by qooob** [comment]
I can not reproduce it on and browser, in any OS.

**Comment by zamolxe** [comment]
Well I'm pretty sure it's not the browser... It's something happening within webmin+authentic theme.
I will try running some stack traces on webmin se wtf is doing

**Comment by zamolxe** [comment]
Do you have any dependencies on any perl modules or anything within webmin ?
What can differ between our environments is that mine is a fresh one, with webmin 1.760 installed.
I've tried removing the theme completely also clean all traces of it from the fs/cache and installed directly the v16... still no go. cleaned again... installed v15.51... that one works fine.

**Comment by qooob** [comment]
That is so wired. Reboot the system!! I think personally, it's stuckness of Webmin cache somewhere..

**Comment by qooob** [comment]
Did you try Chrome?

**Comment by zamolxe** [comment]
Tried Firefox, Safari, Chrome ( Mac OS X )
v15.51 works fine
v16.00 redirect loop
I suspect some dependency to something which was left around in webmin before 1.760, because I installed directly the latest version... otherwise, don't know how big are the changes between 15.51 and 16.

**Comment by qooob** [comment]
Changes are huge! Installed directly? Using `Webmin->Webmin Configuration->Themes` and hen install theme?

**Comment by zamolxe** [comment]
Yes

**Comment by qooob** [comment]
I don't know why it's happening. Will see what others will say on testing! Please wait..

**Comment by andreychek** [comment]
I don't imagine you have a machine experiencing this issue, that we could remotely access and log in to?  A test system, perhaps?

**Comment by zamolxe** [comment]
@qooob 
You are not alone there :)
You can try to reproduce my environment:
Clean CentOS7 install (up to date), then install the webmin-1.760 rpm... 
That should come with v14 of your theme... Just upgrade to the latest one...
I think you will get the same results like me. 
I can also provide the list of rpm's installed.

@andreychek 
I can bring one machine up for testing and grant access to that one.. (if I will reproduce the issue)

**Comment by qooob** [comment]
@zamolxe what is the output of `ls -lsaZ /usr/libexec/webmin/authentic-theme` and `ls -lsaZ /etc/webmin/authentic-theme`, post it please!?

**Comment by zamolxe** [comment]
You have it below... selinux is disable on my end.
# ls -lsaZ /usr/libexec/webmin/authentic-theme /etc/webmin/authentic-theme/
/etc/webmin/authentic-theme/:
total 12
drwxr-xr-x root root ?                                .
drwxr-xr-x root root ?                                ..
-rw-r--r-- root root ?                                config

/usr/libexec/webmin/authentic-theme:
total 300
drwxr-xr-x root root ?                                .
drwxr-xr-x root root ?                                ..
-rwxr-xr-x root root ?                                authentic-lib.cgi
-rwxr-xr-x root root ?                                authentic.pl
-rwxr-xr-x root root ?                                buttons.cgi
-rw-r--r-- root root ?                                CHANGELOG.md
-rwxr-xr-x root root ?                                changelog.pl
-rw-r--r-- root root ?                                config
drwxr-xr-x root root ?                                images
-rwxr-xr-x root root ?                                index.cgi
drwxr-xr-x root root ?                                lang
drwxrwxr-x root root ?                                lib
-rw-r--r-- root root ?                                LICENSE
-rwxr-xr-x root root ?                                navigation.cgi
-rw-r--r-- root root ?                                README.md
-rwxr-xr-x root root ?                                session_login.cgi
-rwxr-xr-x root root ?                                settings.cgi
-rwxr-xr-x root root ?                                settings-editor_read.cgi
-rwxr-xr-x root root ?                                settings-editor_write.cgi
-rwxr-xr-x root root ?                                settings-favorites_save.cgi
-rwxr-xr-x root root ?                                settings-upload.cgi
-rwxr-xr-x root root ?                                settings-upload_save.cgi
-rwxr-xr-x root root ?                                sysinfo.cgi
-rw-r--r-- root root ?                                theme.info
drwxr-xr-x root root ?                                unauthenticated
-rw-r--r-- root root ?                                VERSION.txt

**Comment by qooob** [comment]
Everything looks alright!

**Comment by zamolxe** [comment]
I've got some updates: Trying to query the root (/) ends up with the loop, but if I try to access only a module, in my case the dhcpd configuration (/dhcpd/) that works just fine.

**Comment by qooob** [comment]
What do you mean by `Trying to query the root`? How do you query it?

**Comment by qooob** [comment]
ahh.. looking..

**Comment by qooob** [comment]
Ok, can you please find the line 2008 in `authentic-lib.cgi` and add after `&header($title)` the following:

```Perl
prt($ENV{'REQUEST_URI'})
```

Now, please go to `sub init` and completely comment out/remove this:

```Perl
# Redirect user away, in case requested mode can not be satisfied
    if (   $ENV{'REQUEST_URI'} ne '/'
        && $ENV{'REQUEST_URI'} ne '/?virtualmin'
        && $ENV{'REQUEST_URI'} ne '/?cloudmin'
        && $ENV{'REQUEST_URI'} ne '/?mail'
        && $ENV{'REQUEST_URI'} ne '/?dashboard'
        && index( $ENV{'REQUEST_URI'}, 'xhr' ) lt 0 )
    {
        my $webmin
            = ( $ENV{'HTTPS'} ? 'https://' : 'http://' )
            . $ENV{'HTTP_HOST'} . '/';
        print "Location: $webmin\n\n";
    }
    elsif (
        ( $t_uri_virtualmin != -1 && !&foreign_available("virtual-server") )
        || ( $t_uri_cloudmin != -1 && !&foreign_available("server-manager") )
        || ($t_uri_webmail != -1
            && (&get_product_name() ne 'usermin'
                || ( &get_product_name() eq 'usermin'
                    && !&foreign_available("mailbox") )
            )
        )
        )
    {
        print "Set-Cookie: redirect=0; path=/\r\n";
        my $webmin
            = ( $ENV{'HTTPS'} ? 'https://' : 'http://' )
            . $ENV{'HTTP_HOST'} . '/';
        print "Location: $webmin\n\n";
    }
```

Afterwards, reload the page. What is the output of the printed `$VAR1` that we added to the line `2009`?

**Comment by zamolxe** [comment]
Are you sure is prt($ENV{'REQUEST_URI'}) and not print ?

**Comment by qooob** [comment]
Yes, I got that manual function there..

**Comment by qooob** [comment]
It's better to find things..

**Comment by zamolxe** [comment]
HTTP/1.0 500 Perl execution failed Server: MiniServ/1.760 Date: Sun, 20 Sep 2015 11:53:28 GMT Content-type: text/html; Charset=iso-8859-1 Connection: close
Error - Perl execution failed

Undefined subroutine &main::prt called at /usr/libexec/webmin/authentic-theme/authentic-lib.cgi line 2009.

**Comment by qooob** [comment]
Like I said :) It's cache issue! In the end of the file there should be

```Perl
sub prt {
    my ($____v) = @_;
    use Data::Dumper;
    print '<div style="color: red">';
    print Dumper $____v;
    print '</div>';
}
```

Do you have this in your file? Did you really restart Webmin? Did you restart machines? (please say as it is, I need it for the future reference)

You could get error about not having `use Data::Dumper`.

Take a look at this line:
https://github.com/qooob/authentic-theme/blob/master/authentic-lib.cgi#L2358


OK instead of `prt($ENV{'REQUEST_URI'}) ` do this:


print '++++' . $ENV{'REQUEST_URI'} . '++++';

What is the output?

**Comment by qooob** [comment]
In case you have `/` from `$ENV{'REQUEST_URI'}`, just manually update your .lib file and restart Webmin.

**Comment by zamolxe** [comment]
If I query https://1.2.3.4:10000 the output is / and the loop is gone.

**Comment by qooob** [comment]
But what do you query on regular basis?

**Comment by qooob** [comment]
This is what you should do?

**Comment by zamolxe** [comment]
Usually I go directly yo https://1.2.3.4:10000/ and sometimes directly to https://1.2.3.4:10000/dhcpd
There is nothing fancy about the queries

**Comment by qooob** [comment]
Second will never work!! 

What do you get on the print if you query the first example: `https://1.2.3.4:10000/` ?

**Comment by qooob** [comment]
Actually, if I query https://1.2.3.4:10000/fsdump/ it works without left menu. I wouldn't recommend doing this. Why not just query `https://1.2.3.4:10000`... However `https://1.2.3.4:10000/` works also just fine in my case.

**Comment by zamolxe** [comment]
Ok.. I found out the root cause... 
After installing perl-Data-Dumper to get your function working... Everything worked without any issues.
Just to confirm, I removed perl-Data-Dumper... got back the loop...
Can someone else confirm this ?

**Comment by qooob** [comment]
Reallly??? Let me check. This is only possible if I forgot to remove it somewhere - that would be bad!

**Comment by qooob** [comment]
Checking...

**Comment by zamolxe** [comment]
Fix for it is following:
replace the use with eval in front...

sub prt {
    my ($____v) = @_;
    eval "use Data::Dumper";
    print '<div style="color: red">';
    print Dumper $____v;
    print '</div>';
}

**Comment by qooob** [comment]
But this function is not intended to run on users.. and I just checked it - there is not a single place where I would call it..

**Comment by zamolxe** [comment]
I've seen... Don't know what is actually going on wrong in there, but like I said...
Installing that package made it work all the time... removing it got the loop back.
Patching your prt function, fixed the issue without having to install the dumper rpm again.

You can check to see if you can reproduce the loop by removing the dumper from your system.

**Comment by qooob** [comment]
Listen, thank you so much, pal!! I wouldn't believe this unless I tried to simply renaming it something else. I got redirect loop. You're right.. Fixing... Thanks!!!

**Comment by qooob** [comment]
:)

**Comment by zamolxe** [comment]
Glad to give something back to the community! 
Thanks for a nice troubleshooting weekend :)

**Comment by qooob** [comment]
...but do you realize that this function also in all previous versions? How come it's alright with those versions and fails with 16. Any thoughts? ;)

**Comment by zamolxe** [comment]
To be honest with you... I have no idea why it brakes. :|

**Comment by qooob** [comment]
I see. Alright! Thank you anyway!! ;)

**Comment by spye** [comment]
Just coming in to this thread... I just wanted to point out that I am having exactly the same issue as above, with Webmin 1.770 and Authentic 16. Tried all the same steps before coming to this page: clearing local cache, server cache, restarting webmin, rebooting server, changing webmin to non-SSL, switching ports... tried it all, same issue. Also running CentOS 7.

**Comment by spye** [comment]
And.... I just found your v17 update. So here's what I did to fix the bug: I edited /etc/webmin/config, and changed referrers_none=1 to 0. Then I manually added /webmin on the URL after logging in (which is when I get the redirect loop), and that gave me access to the webmin config, where I could then go and manually update the theme to the version available from this repo. I had this happen on several CentOS 7 systems with this same configuration, so it was definitely reproducible for me, but this fix resolved it completely. Odd though, that v17 of your theme doesn't appear to address bug #247, so I'm guessing I missed an update somewhere that caused it to break in the first place.
